File: software/examples/rust/example_threshold.rs
use std::{error::Error, io, thread, time::Duration};
use tinkerforge::{ip_connection::IpConnection, laser_range_finder_bricklet::*};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Laser Range Finder Bricklet.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let lrf = LaserRangeFinderBricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    // Turn laser on and wait 250ms for very first measurement to be ready
    lrf.enable_laser().recv()?;
    thread::sleep(Duration::from_millis(250));

    // Get threshold receivers with a debounce time of 10 seconds (10000ms).
    lrf.set_debounce_period(10000);

    let distance_reached_receiver = lrf.get_distance_reached_callback_receiver();

    // Spawn thread to handle received callback messages.
    // This thread ends when the `lrf` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for distance_reached in distance_reached_receiver {
            println!("Distance: {} cm", distance_reached);
        }
    });

    // Configure threshold for distance "greater than 20 cm".
    lrf.set_distance_callback_threshold('>', 20, 0);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;

    lrf.disable_laser().recv()?; // Turn laser off

    ipcon.disconnect();
    Ok(())
}

File: software/examples/go/example_threshold.go
package main

import (
	"fmt"
	"time"
	"tinkerforge/ipconnection"
	"tinkerforge/laser_range_finder_bricklet"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your Laser Range Finder Bricklet.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	lrf, _ := laser_range_finder_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	// Turn laser on and wait 250ms for very first measurement to be ready
	lrf.EnableLaser()
	time.Sleep(250 * time.Millisecond)

	// Get threshold receivers with a debounce time of 10 seconds (10000ms).
	lrf.SetDebouncePeriod(10000)

	lrf.RegisterDistanceReachedCallback(func(distance uint16) {
		fmt.Printf("Distance: %d cm\n", distance)
	})

	// Configure threshold for distance "greater than 20 cm".
	lrf.SetDistanceCallbackThreshold('>', 20, 0)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()

	lrf.DisableLaser() // Turn laser off
}
